addon configuration enhancement (#58)
open-cluster-management-io/enhancements: enhancements/sig-architecture/58-addon-placement-annotation/README.md

# Setting Up Addon Configuration 

## Release Signoff Checklist

- [] Enhancement is `implemented`
- [] Design details are appropriately documented from clear requirements
- [] Test plan is defined
- [] Graduation criteria for dev preview, tech preview, GA
- [] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)

## Summary

The proposal introduces some ways to help user to set up configuration for their addons.

## Motivation

We want to provide a way to help user to set up configuration for their addons with current addon APIs

## Use cases

> 1. As a user, I have a configuration for my all managed cluster addons, the addons share the same configuration.

Reviewer: I think we should have some details here.

Author: use the specific use case

	// This field should be configured if the add-on have a CRD that controls the configuration of the add-on.
	// +optional
	CRDName string `json:"crdName"`

	// crName is the name of the CR used to configure instances of the managed add-on.
	// This field should be configured if add-on CR have a consistent name across the all of the ManagedCluster instaces.
	// +optional
	CRName string `json:"crName"`

	// lastObservedGeneration is the observed generation of the custom resource for the configuration of the addon.
	// +optional
	LastObservedGeneration int64 `json:"lastObservedGeneration,omitempty"`
}
```

but the `ConfigCoordinates` has some limits, it only support to use `CustomResourceDefinition` as the addon
configuration, and for addon developers, it's not easy to locate the configuration object in their controllers. we want
to refactor the `ConfigCoordinates` to

``` golang
// ConfigCoordinates contains enough information to let one addon identify its configuration object.
// A config object must be in the same namespace with its addon, or be cluster-scoped, so there is no
// namespace field.
type ConfigCoordinates struct {
	// API version of the referent.

> 	APIVersion string `json:"apiVersion"`

Reviewer: let's use gvr here. and also default namespace/name??

Author: use the gvr and add the default ns and name

	Kind string `json:"kind"`
	// Name of the referent.
	// More info: http://kubernetes.io/docs/user-guide/identifiers#names
	Name string `json:"name"`
}
```

with this change, the `ConfigCoordinates` will be able to support all types of configuration, and the developers will
be able to get the `GroupVersionResource` of the configuration object.

In the `ManagedClusterAddOn` API, we will also introduce `AddOnConfiguration` option, When addon has its own
configuration, the addon can use this option to locate its configuration

```golang
// ManagedClusterAddOnSpec defines the install configuration of
// an addon agent on managed cluster.
type ManagedClusterAddOnSpec struct {
	// installNamespace is the namespace on the managed cluster to install the addon agent.
	// If it is not set, open-cluster-management-agent-addon namespace is used to install the addon agent.
	// +optional
	// +kubebuilder:validation:MaxLength=63
	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
	InstallNamespace string `json:"installNamespace,omitempty"`

	// addOnConfiguration is a configuration object reference that is used by current addon.
	// +optional

> 	AddOnConfiguration ConfigCoordinates `json:"addOnConfiguration,omitempty"`

Reviewer: we should only have namespace/name here. Also we need update status field

Author: yes, we need update status

}
```

As a specific case, sometimes user want to assign their addons on specific nodes, we want to introduce a `AddOnDeployementConfig`
to support this.

```golang
// AddOnDeployementConfig represents a deployment configuration for an add-on.
// Currently, user can use this to configure nodeSelector, toleration and variables.

> type AddOnDeployementConfig struct {

Reviewer: namespace/cluster scoped?

Author: it is a cluster scope

	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec represents a desired configuration for an add-on.
	// +required
	Spec AddOnDeployementConfigSpec `json:"spec"`
}

type AddOnDeployementConfigSpec struct {
	// CustomizedVariables is a list of name-value variables for the current add-on.
	// add-on can use these variables to render its deployment.
	// The default is an empty list.
	// +optional
	CustomizedVariables []CustomizedVariable `json:"customizedVariable,omitempty"`

	// NodeSelector defines which Nodes the Pods are scheduled on.
	// The default is an empty list.
	// +optional

> 	NodeSelector map[string]string `json:"nodeSelector,omitempty"`

Reviewer: I think we should put it in a structure working as a duck type maybe. Some thing like nodeSchedulable

Author: separate to a struct

# The Placement Annotation of Addon

## Release Signoff Checklist

- [] Enhancement is `implemented`
- [] Design details are appropriately documented from clear requirements
- [] Test plan is defined
- [] Graduation criteria for dev preview, tech preview, GA
- [] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)

## Summary

The proposal introduces two annotations to help user to specify node selector and tolerations for their addons.

## Motivation

> Sometimes user want to place their addons on specific nodes, so we need a way to set the node selector and tolerations for a addon.

Reviewer: a --> an

Author: fixed

In this proposal, if one addon has the node selector and tolerations annotations, this function will merge them to the `Values` object with the key `NodeSelector` and `Tolerations`.

For a addon implementation, developers use [`AgentAddonFactory`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L28) to build their addon, they use [`WithGetValuesFuncs`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L61) to specify the get values functions (the function type is (`GetValuesFunc`)[https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L24]) to render their addon deployment, for get values function, the developers can choose to use addon-frameworks `GetValuesFromAddonAnnotation` function or implement it by themselves to get these two annotations.

### Test Plan
- test add placement annotation to a addon which has no values annotation 
- test add placement annotation to a addon which has values annotation 
- test delete placement annotation from a addon

### Graduation Criteria
N/A

### Upgrade Strategy

> N/A

Reviewer: any concern on the upgrade?

Author: add a probable upgrade case, and expecting other reviewers give more comments

- [] Graduation criteria for dev preview, tech preview, GA
- [] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)

## Summary

The proposal introduces two annotations to help user to specify node selector and tolerations for their addons.

## Motivation

Sometimes user want to place their addons on specific nodes, so we need a way to set the node selector and tolerations for an addon.

## Proposal

We introduce two annotations to set node selector and tolerations for an addon

1. Node selector annotation, the key of this annotation is: `addon.open-cluster-management.io/nodeSelector`, the value of this annotation represents a node selector with json format
2. Tolerations annotation, the key of this annotation is: `addon.open-cluster-management.io/tolerations`, the value of this annotation represents a toleration array with json format

There is an example

```yaml
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: ManagedClusterAddOn
metadata:
  name: work-manager
  namespace: cluster1
  annotations:
    addon.open-cluster-management.io/nodeSelector: "{\"kubernetes.io/os\":\"linux\"}"

>     addon.open-cluster-management.io/tolerations: "[{\"key\":\"node.kubernetes.io/unreachable\",\"operator\":\"Exists\",\"effect\":\"NoSchedule\"}]"

Reviewer: I think values is reserved for other usage? what if a user or another controller set the value in advance. I think it will be hard to merge them.

>     addon.open-cluster-management.io/tolerations: "[{\"key\":\"node.kubernetes.io/unreachable\",\"operator\":\"Exists\",\"effect\":\"NoSchedule\"}]"

Reviewer 2: I just use values as an example, we could find a new name if using one annotation is accepted.

>     addon.open-cluster-management.io/tolerations: "[{\"key\":\"node.kubernetes.io/unreachable\",\"operator\":\"Exists\",\"effect\":\"NoSchedule\"}]"

Author: I think use two independent annotations is more clear and more easier to maintain

> # The Placement Annotation of Addon

Reviewer 3: Minor nit. Can we use a different word instead of `Placement`? We have the Placement API and this enhancement is talking about something that is not related to that.

Author: change to `The Addon Node Selector and Toleration Annotation`

## Release Signoff Checklist

- [] Enhancement is `implemented`
- [] Design details are appropriately documented from clear requirements
- [] Test plan is defined
- [] Graduation criteria for dev preview, tech preview, GA
- [] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)

## Summary

The proposal introduces two annotations to help user to specify node selector and tolerations for their addons.

## Motivation

> Sometimes user want to place their addons on specific nodes, so we need a way to set the node selector and tolerations

Reviewer 3: ```suggestion
Sometimes user want to assign their addons on specific nodes, so we need a way to set the node selector and toleration
```

for an addon.

## Proposal

> We introduce two annotations to set node selector and tolerations for an addon

Reviewer 3: ```suggestion
We introduce two annotations to set node selector and toleration for an addon:
```

1. Node selector annotation, the key of this annotation is: `addon.open-cluster-management.io/nodeSelector`, the value of this annotation represents a node selector with json format
2. Tolerations annotation, the key of this annotation is: `addon.open-cluster-management.io/tolerations`, the value of this annotation represents a toleration array with json format

> There is an example

Reviewer 3: ```suggestion
Example:
```

### Test Plan
- test add placement annotation to an addon which has no values annotation 
- test add placement annotation to an addon which has values annotation
- test delete placement annotation from an addon which has no values annotation
- test delete placement annotation from an addon which has values annotation

### Graduation Criteria
N/A

### Upgrade Strategy

If there is an existing addon CR that already used the addon values annotation to specify the node selector and tolerations and the addon implementation used the `GetValuesFromAddonAnnotation` function to get them. The addon user can choose to keep them in addon values annotation or use new addon placement annotation to define them and remove them from the addon values annotation.

### Version Skew Strategy
N/A

> ## Alternatives

Author: yeah, we have considered this, if we add these to addon spec, it may not meet all needs, e.g. if one addon need to deploy more than one agent, and these agents have different node selector/tolerations, this will be hard to use

to keep the addon flexibility, we proposed this, providing an optional way for the addon developers, they can choose use this way or use their own way to handle the node selector/tolerations for their addons, e.g. the developers may have own addon configuration cr, and use the configuration to render their addon agents

> ## Alternatives

Reviewer: I think we cannot ensure the agent will use this even if adding this to spec. installNamespace is different, since it is needed for the registration controller to put secret for the addon to communicate to hub. @skeeey let's add spec as an atlernative in the proposal, and clarify the reason.

> ## Alternatives

Author: added

> ## Alternatives

Reviewer 4: I had the same thought. I think the additional flexibility is a negative in this case. We want a consistent interface for how to configure addons from a user's perspective. If this is free-form JSON strings that are different per addon, I think this becomes hard to use and confusing. It's especially confusing when a user puts in an invalid JSON string since there is no immediate feedback from API validation. It also makes API compatibility hard since it becomes up to each OCM addon how to interpret that JSON strin.

I also dislike that an annotation has a JSON string as a value since these are not user-friendly to read or write and suggest that a new item in the spec is more appropriate.

My suggestion is to make the spec flexible enough for all current OCM addons and make it optional. Any complex configurations beyond this can be addon specific in their own CRDs.

> ## Alternatives

Reviewer: but nodeSelect/toleration is not a common property for every addon. Having nodeSelect/toleration in the spec gives an impression that the the addon will honor this field, but it is not true. If we want a more strict validation, I think we should have an Object ref field (@jnpacker mentioned this also) instead of putting nodeselector in the addon spec.

> ## Alternatives

Reviewer 5: What about an objectRef or list of objectRef's.  It does mean that each addon may not configure the same set of values or even resources, but if as @mprahl  suggests, you used a CRD for your addon, you could have basic validation, defaults, etc...

Also a ConfigMap or secret could be leveraged to try things out, and at least the format is well understood.

> ## Alternatives

Author: I will add the objectRef to the enhancement as an optional proposal

> ## Alternatives

Reviewer 4: @qiujian16 and @jnpacker this sounds interesting. I hadn't heard about `objectRef` before and couldn't find anything concrete online. Could you please explain what this concept is?

> ## Alternatives

Reviewer 3: @mprahl probably this https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5622-L5639
